android/app/src/main/java/com/criticalblue/auth/demo/auth/AuthRepo.java

```diff
@@ -61,45 +63,92 @@
+
     String clientId;
     String redirectUri;
     String authScope;
 
-    ConnectionBuilder connectionBuilder;
-
     public AuthRepo(BooksApp app) {
         this.app = app;
 
         loginLock = new Semaphore(1);
 
         loginListener = null;
-        logoutListener = null;
-
-        connectionBuilder = new SignedConnectionBuilder(app.getAssets());
-
-        AppAuthConfiguration.Builder builder = new AppAuthConfiguration.Builder();
-        builder.setConnectionBuilder(connectionBuilder);
-        authService = new AuthorizationService(app, builder.build());
         authState = null;
         userInfoUrl = null;
         clientId = null;
         redirectUri = null;
         authScope = null;
+
+        authService = new AuthorizationService(app, ApproovCustomConnection());
+    }
+
+    /**
+     * All the calls made by AppAuth will be secured by using the Approov Managed Trust Roots
+     * feature to protect the TLS channel against MitM attacks.
+     *
+     * @link https://approov.io/docs/latest/approov-usage-documentation/#managed-trust-roots
+     *
+     * @return The AppAuthConfiguration with a HttpsURLConnection secured by Approov
+     */
+    private AppAuthConfiguration ApproovCustomConnection() {
+        return new AppAuthConfiguration.Builder()
+                .setConnectionBuilder(new ConnectionBuilder() {
+                    @NonNull
+                    @Override
+                    public HttpsURLConnection openConnection(@NonNull Uri uri) throws IOException {
+                        URL url = new URL(uri.toString());
+                        HttpsURLConnection connection = (HttpsURLConnection) url.openConnection();
+
+                        io.approov.service.httpsurlconn.ApproovService.addApproov(connection);
+
+                        return connection;
+                    }
+                })
+                .build();
     }
 
+    /**
+     * Approov allows to fetch secrets just in time for being used to make API calls, provided that
+     * the mobile app passes the attestation, thus no more hardcoded secrets on the app code.
+     *
+     * @link https://approov.io/docs/latest/approov-usage-documentation/#secure-strings
+     *
+     * @return The client secret retrieved from the Approov Cloud service
+     */
+    private String fetchJustInTimeClientSecret() {
+        String clientSecret = null;
+
+        try {
+            clientSecret = io.approov.service.httpsurlconn.ApproovService.fetchSecureString("client_secret", null);
+        } catch (ApproovRejectionException e) {
+            // failure due to the attestation being rejected, e.getARC() and e.getRejectionReasons() may be used to present information to the user
+            // (note e.getRejectionReasons() is only available if the feature is enabled, otherwise it is always an empty string)
+            failLogin(new AuthException("Failed to fetch the client_secret. ARC: " + e.getARC() + "REASON: " + e.getRejectionReasons()));
```

Comment:
Have you tried to see what happens if this fails. Do the messages appear to the user. They are a bit cryptic from a user perspective if so, this one should be able the "app or device failing verification checks or somesuch"

Reply:
I used the messages because this a demo app targetting devs, thus I though to be useful to see the what went wrong on the screen.

I understand that this may encourage the wrong use of them, thus I will move it to logging and replace with a message in the likes you suggested.